src/renderers/WebGLRenderer.js

```diff
@@ -1582,7 +1583,7 @@ class WebGLRenderer {
 
 					needsProgramChange = true;
 
-				} else if ( material.fog === true && materialProperties.fog !== fog ) {
+				} else if ( material.fog === true && ( materialProperties.fog !== fog || ( fog && ( materialProperties.fogSquared !== fog.squared ) ) ) ) {
```

Comment:
This and line 1397 seems to work, but I am not 100% confident that all conceivable edge cases are perfectly covered.

Reply:
Alternatives to checking `squared` every time:
1. Make it truly or practically immutable after construction. I don't know what pattern would be preferred here.
2. Have separate fog classes for FOG_EXP and FOG_EXP2 (`DensityFogPhysical` and `DensityFogSquared` or whatever, still kind of clunky). Independent default densities would be a bonus.

One could also have an abstract parent class. Sadly, the name "Fog" will still be in use, but maybe `SceneFog`? With inheritable `color`, `isSceneFog`, and default shader chunks, and extensible with custom shader chunks, instead of juggling with the `#define`s in the renderer and separate shader chunk files. The proposed three standard fogs could be implemented using this, and it would also be straightforward to, e.g., make a fog with angle-dependent density, combine density with range, sneak in a fog texture, etc.. (If the default chunks match RangeFog, then the name `Fog` can even be reused, and the "abstract" class will be concrete and backwards compatible. In that case, I would suggest using the more "traditional" names `FogExp` and `FogExp2` on the subclasses.)

Reply:
> Make it truly or practically immutable after construction. I don't know what pattern would be preferred here.

Is there a use case in which it would be updated after construction? We can just say that it shouldn't be changed.

Reply:
I update it after construction in the example. That is how I encountered the updating issue. It seems to work, so I guess there is no urgent need to make it _less_ user friendly. The `color` and `density` are uniforms that can be changed after construction, and as a user, I would expect the same flexibility for `squared`.